src/data-structures/Bytes.huff

```diff
@@ -0,0 +1,114 @@
+    // TODO: optimize SWAPS
+
+    dup3 0x20 sub                     // [pad_len, word, is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+    // swap1 swap2 0x20 sub              // [pad_len, word, is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+    0x08 mul swap1 dup2               // [shift, word, shift, is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+    shr                               // [left_padded_word, shift, is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+    swap1 shl                         // [clean_word, is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+    dup5 mstore                       // [is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+    swap3 add swap2                   // [is_sec_loop, index(i), index(j + 1), free_loc_pos]
+  
+    // here we check if current loop is for the 2nd array
+    swap1 pop                         // [is_sec_loop, index(j + 1), free_loc_pos]
+    iszero bridge jumpi               // [index(j + 1), free_loc_pos]
+    pop break jump                    // [free_loc_pos]
+
+    empty_slot:                       // [is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+      swap2 pop pop                   // [is_sec_loop, index(j), free_loc_pos]
+      iszero bridge jumpi             // [index(j), free_loc_pos]  
+      pop break jump                  // [free_loc_pos]
+  
+    bridge:                           // [index(j), free_loc_pos]
+      <mem_ptr2> 0x20 add             // [index(i), index(j), free_loc_pos]
+      <mem_ptr2> mload                // [len2, index(i), index(j), free_loc_pos]
+      0x01                            // [is_sec_loop, len2, index(i), index(j), free_loc_pos]
+      loop jump
+        
+    full_slot:                        // [word, is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+      dup5 mstore                     // [is_sec_loop, len_left, index(i), index(j), free_loc_pos]
+      swap1 0x20 swap1 sub            // [len_left - 0x20, is_sec_loop, index(i), index(j), free_loc_pos]
+      swap2 0x20 add                  // [index(i + 1), is_sec_loop, len_left - 0x20, index(j), free_loc_pos]
+      swap3 0x20 add                  // [index(j + 1), is_sec_loop, len_left - 0x20, index(i + 1), free_loc_pos]
+      swap3 swap2 swap1               // [is_sec_loop, len_left - 0x20, index(i + 1), index(j + 1), free_loc_pos]
+      loop jump
+  
+  break:                // [free_loc_pos]
+    
+}
+
+// @param mem_ptr position in memory of the start of the bytes array 
+// @param start position of the slice relative to the array
+// @param length of the output slice
+// @returns position of the output slice in memory
+#define macro SLICE_MEMORY(mem_ptr, start, length) = takes(3) returns(1) {
+  <mem_ptr> mload              // [len]  
+  
+  // TODO: assert that length <= len
+  // TODO: assert that start < len + mem_ptr
```

Comment:
Re: `assert that length <= len` and `start < len + mem_ptr`, in a high level language i would say assert both, but since this is a lower level language it should be up to the developer to ensure this is the case.

However, I would document very clearly what assumptions we make.